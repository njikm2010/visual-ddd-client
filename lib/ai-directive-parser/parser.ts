/**
 * 从 GPT 返回的结果中解析出指令
 */
import trimStart from 'lodash/trimStart';
import trimEnd from 'lodash/trimEnd';

const DELIMITER = '%%';
const DIRECTIVE_REG = /%%(.+?)%%/gm;

export interface Directive {
  type: string;
  params: Record<string, any>;
}

export function parseItem(text: string): Directive | null {
  if (!text) {
    return null;
  }

  const sections = trimEnd(trimStart(text, DELIMITER), DELIMITER)
    .split(' ')
    .map(i => i.trim())
    .filter(Boolean);
  const directive = sections.shift();

  if (!directive) {
    return null;
  }

  const params = sections.reduce<Record<string, any>>((acc, item) => {
    const [key, value] = item.split('=');

    const rawValue = trimStart(trimEnd(value, '"'), '"');
    acc[key] = rawValue;

    return acc;
  }, {});

  const dir: Directive = {
    type: directive,
    params,
  };

  return dir;
}

/**
 * 解析指令
 * @param input
 * @returns
 */
export function parse(input: string): Directive[] | null {
  if (!input.includes(DELIMITER)) {
    return null;
  }

  const matched = input.match(DIRECTIVE_REG);

  if (!matched) {
    return null;
  }

  const directives: Directive[] = [];

  for (const item of matched) {
    const directive = parseItem(item);
    if (directive) {
      directives.push(directive);
    }
  }

  return directives.length ? directives : null;
}
