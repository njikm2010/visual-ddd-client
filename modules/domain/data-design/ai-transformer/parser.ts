/**
 * 从 GPT 返回的结果中解析出指令
 */
import trimStart from 'lodash/trimStart';
import trimEnd from 'lodash/trimEnd';
import {
  DirectiveName,
  GroupedTransformDirectives,
  TransformDirective,
  getTableName,
  isAvailableDirective,
  RenameFieldDirective,
  isTableDirective,
  TableFieldReference,
} from './protocol';
import { NoopArray } from '@wakeapp/utils';

const DELIMITER = '%%';
const DIRECTIVE_REG = /%%(.+)%%/gm;

function normalizedValue(value: string) {
  if (value === 'true') {
    return true;
  } else if (value === 'false') {
    return false;
  }

  return value;
}

function parseItem(text: string): TransformDirective | null {
  if (!text) {
    return null;
  }

  const sections = trimEnd(trimStart(text, DELIMITER), DELIMITER)
    .split(' ')
    .map(i => i.trim())
    .filter(Boolean);
  const directive = sections.shift();

  if (!directive) {
    return null;
  }

  if (!isAvailableDirective(directive)) {
    console.warn(`Unknown directive: ${directive}`);
    return null;
  }

  const params = sections.reduce<Record<string, any>>((acc, item) => {
    const [key, value] = item.split('=');

    const rawValue = trimStart(trimEnd(value, '"'), '"');
    acc[key] = normalizedValue(rawValue);

    return acc;
  }, {});

  return {
    type: directive,
    params,
  } as TransformDirective;
}

export function parse(input: string): TransformDirective[] | null {
  if (!input.includes(DELIMITER)) {
    return null;
  }

  const matched = input.match(DIRECTIVE_REG);

  if (!matched) {
    return null;
  }

  const directives: TransformDirective[] = [];

  for (const item of matched) {
    const directive = parseItem(item);
    if (directive) {
      directives.push(directive);
    }
  }

  return directives.length ? directives : null;
}

export function parseReference(id: TableFieldReference) {
  const [table, field] = id
    .trim()
    .split('.')
    .map(i => i.trim());

  return {
    table,
    field,
  };
}

/**
 * 分组指令
 * @param directives
 * @returns
 */
export function groupByTable(directives: TransformDirective[]): GroupedTransformDirectives[] {
  const grouped: Map<string, GroupedTransformDirectives> = new Map();
  /**
   * 被重命名的表格，需要进行合并
   */
  const tableRenamed: Map<string, string> = new Map();

  for (const directive of directives) {
    const tableName = getTableName(directive);
    let group = grouped.get(tableName);

    if (group == null) {
      group = {
        table: tableName,
        tableDirectives: [],
        fieldDirectives: [],
      };
      grouped.set(tableName, group);
    }

    if (isTableDirective(directive)) {
      group.tableDirectives.push(directive);
      if (directive.type === DirectiveName.RenameTable) {
        // 表格进行了重命名, 记录一下
        tableRenamed.set(directive.params.name, directive.params.newName);
      }
    } else {
      group.fieldDirectives.push(directive);
    }
  }

  // 后处理，对重命名的表格进行合并
  if (tableRenamed.size) {
    for (const [name, newName] of tableRenamed.entries()) {
      const groupByName = grouped.get(name);
      const groupByNewName = grouped.get(newName);

      if (groupByNewName == null) {
        continue;
      }

      const newGroup: GroupedTransformDirectives = {
        table: name,
        tableDirectives: [
          ...(groupByName?.tableDirectives ?? NoopArray),
          ...(groupByNewName?.tableDirectives ?? NoopArray),
        ],
        fieldDirectives: [
          ...(groupByName?.fieldDirectives ?? NoopArray),
          ...(groupByNewName?.fieldDirectives ?? NoopArray),
        ],
      };

      grouped.delete(newName);
      grouped.set(name, newGroup);
    }
  }

  // 后处理，对重命名的字段进行规范化
  // 让字段可能被跟踪
  for (const group of grouped.values()) {
    const list = group.fieldDirectives;

    const renameDirectives = list.filter((i): i is RenameFieldDirective => i.type === DirectiveName.RenameField);

    if (!renameDirectives.length) {
      continue;
    }

    // map newName -> name
    const mapper = renameDirectives.reduce<Record<string, string>>((acc, cur) => {
      acc[cur.params.newName] = cur.params.name;

      return acc;
    }, {});

    // 修正
    for (const dir of list) {
      if (dir.type !== DirectiveName.RenameField && dir.params.name in mapper) {
        dir.params.name = mapper[dir.params.name];
      }

      if ((dir.type === DirectiveName.UpdateField || dir.type === DirectiveName.AddField) && dir.params.reference) {
        // 修正引用
        const { table, field } = parseReference(dir.params.reference);
        if (table in mapper) {
          dir.params.reference = `${mapper[table]}.${field}`;
        }
      }
    }
  }

  return Array.from(grouped.values());
}
