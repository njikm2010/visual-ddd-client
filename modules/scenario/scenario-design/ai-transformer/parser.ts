import { toNameCase } from '@/lib/utils';
import { AVAILABLE_DIRECTIVES, DirectiveName, ScenarioDirective } from './protocol';
import { parse as parseDirectives } from '@/lib/ai-directive-parser';

function isAvailableDirective(name: string): name is DirectiveName {
  return AVAILABLE_DIRECTIVES.includes(name as any);
}

export function parse(input: string): ScenarioDirective[] | null {
  const results = parseDirectives(input);

  if (results == null) {
    return null;
  }

  const directives: ScenarioDirective[] = [];
  for (const i of results) {
    if (!isAvailableDirective(i.type)) {
      continue;
    }

    if ('x' in i.params) {
      i.params.x = parseInt(i.params.x, 10);
    }

    if ('y' in i.params) {
      i.params.y = parseInt(i.params.y, 10);
    }

    if ('name' in i.params) {
      i.params.name = toNameCase('CamelCase', i.params.name);
    }

    if ('from' in i.params) {
      i.params.from = toNameCase('CamelCase', i.params.from);
    }

    if ('to' in i.params) {
      i.params.to = toNameCase('CamelCase', i.params.to);
    }

    directives.push(i as ScenarioDirective);
  }

  return directives;
}
