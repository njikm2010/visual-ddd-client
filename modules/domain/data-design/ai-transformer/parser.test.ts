import { parse, groupByTable } from './parser';

test('parse', () => {
  expect(parse('%%')).toBeNull();
  expect(parse('%%%%')).toBeNull();
  expect(parse('%% %%')).toBeNull();

  expect(parse('%%a%%')).toBeNull();

  expect(parse(` %%renameField table="MyUser" name="name" newName="oName" %% `)).toEqual([
    {
      params: {
        name: 'name',
        newName: 'oName',
        table: 'MyUser',
      },
      type: 'renameField',
    },
  ]);

  expect(
    parse(`
%%addField table="MyUser" name="createdAt" title="创建时间" type="DateTime" notNull="true"%%
%%addField table="MyUser" name="updatedAt" title="更新时间" type="DateTime" notNull="true"%%

%%addField table="Address" name="createdAt" title="创建时间" type="DateTime" notNull="true"%%
%%addField table="Address" name="updatedAt" title="更新时间" type="DateTime" notNull="true"%%
	`)
  ).toEqual([
    {
      params: {
        name: 'createdAt',
        notNull: true,
        table: 'MyUser',
        title: '创建时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
    {
      params: {
        name: 'updatedAt',
        notNull: true,
        table: 'MyUser',
        title: '更新时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
    {
      params: {
        name: 'createdAt',
        notNull: true,
        table: 'Address',
        title: '创建时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
    {
      params: {
        name: 'updatedAt',
        notNull: true,
        table: 'Address',
        title: '更新时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
  ]);
});

test('table name normalized', () => {
  expect(
    parse(`
%%createTable name="myUser" title="用户"%%
%%addField table="MyUser" name="CreatedAt" title="创建时间" type="DateTime" notNull="true"%%
%%addField table="MyUser" name="updated_at" title="更新时间" type="DateTime" notNull="true"%%
%%renameField table="MyUser" name="hello" newName="OK"%%
%%renameTable name="MyUser" newName="account" %%
  `)
  ).toEqual([
    {
      params: {
        name: 'MyUser',
        title: '用户',
      },
      type: 'createTable',
    },
    {
      params: {
        name: 'createdAt',
        notNull: true,
        table: 'MyUser',
        title: '创建时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
    {
      params: {
        name: 'updatedAt',
        notNull: true,
        table: 'MyUser',
        title: '更新时间',
        type: 'DateTime',
      },
      type: 'addField',
    },
    {
      params: {
        name: 'hello',
        newName: 'ok',
        table: 'MyUser',
      },
      type: 'renameField',
    },
    {
      params: {
        name: 'MyUser',
        newName: 'Account',
      },
      type: 'renameTable',
    },
  ]);
});

describe('groupByTable', () => {
  test('simple', () => {
    expect(
      groupByTable(
        parse(`
%%createTable name="Product" title="商品"%%
%%createTable name="Order" title="订单"%%

%%addField table="Product" name="id" title="商品id" type="Long" primaryKey="true" notNull="true"%%
%%addField table="Product" name="name" title="商品名" type="String" notNull="true"%%
%%addField table="Product" name="price" title="价格" type="Decimal" notNull="true"%%
%%addField table="Product" name="description" title="描述" type="LongText" notNull="true"%%

%%addField table="Order" name="id" title="订单id" type="Long" primaryKey="true" notNull="true"%%
%%addField table="Order" name="productId" title="商品引用" type="Reference" reference="Product.id" referenceCardinality="ManyToOne" notNull="true"%%
%%addField table="Order" name="quantity" title="数量" type="Integer" notNull="true"%%
%%addField table="Order" name="totalPrice" title="总价" type="Decimal" notNull="true"%%

  `)!
      )
    ).toMatchSnapshot();
  });

  test('table renamed', () => {
    expect(
      groupByTable(
        parse(`
%%renameTable name="Product" newName="Good"%%
%%createTable name="Order" title="订单"%%

%%addField table="Product" name="id" title="商品id" type="Long" primaryKey="true" notNull="true"%%
%%addField table="Product" name="name" title="商品名" type="String" notNull="true"%%
%%addField table="Good" name="price" title="价格" type="Decimal" notNull="true"%%
%%addField table="Good" name="description" title="描述" type="LongText" notNull="true"%%

%%addField table="Order" name="id" title="订单id" type="Long" primaryKey="true" notNull="true"%%
%%renameField table="Order" name="id" newName="orderNo" %%
%%addField table="Order" name="productId" title="商品引用" type="Reference" reference="Product.id" referenceCardinality="ManyToOne" notNull="true"%%
%%addField table="Order" name="quantity" title="数量" type="Integer" notNull="true"%%
%%addField table="Order" name="totalPrice" title="总价" type="Decimal" notNull="true"%%
%%updateField table="Order" name="id" title="订单编号" %%
%%updateField table="Order" name="orderNo" type="String" %%

  `)!
      )
    ).toMatchSnapshot();
  });
});
