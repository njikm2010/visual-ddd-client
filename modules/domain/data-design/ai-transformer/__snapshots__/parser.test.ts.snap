// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupByTable simple 1`] = `
[
  {
    "fieldDirectives": [
      {
        "params": {
          "name": "id",
          "notNull": true,
          "primaryKey": true,
          "table": "Product",
          "title": "商品id",
          "type": "Long",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "name",
          "notNull": true,
          "table": "Product",
          "title": "商品名",
          "type": "String",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "price",
          "notNull": true,
          "table": "Product",
          "title": "价格",
          "type": "Decimal",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "description",
          "notNull": true,
          "table": "Product",
          "title": "描述",
          "type": "LongText",
        },
        "type": "addField",
      },
    ],
    "table": "Product",
    "tableDirectives": [
      {
        "params": {
          "name": "Product",
          "title": "商品",
        },
        "type": "createTable",
      },
    ],
  },
  {
    "fieldDirectives": [
      {
        "params": {
          "name": "id",
          "notNull": true,
          "primaryKey": true,
          "table": "Order",
          "title": "订单id",
          "type": "Long",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "productId",
          "notNull": true,
          "reference": "Product.id",
          "referenceCardinality": "ManyToOne",
          "table": "Order",
          "title": "商品引用",
          "type": "Reference",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "quantity",
          "notNull": true,
          "table": "Order",
          "title": "数量",
          "type": "Integer",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "totalPrice",
          "notNull": true,
          "table": "Order",
          "title": "总价",
          "type": "Decimal",
        },
        "type": "addField",
      },
    ],
    "table": "Order",
    "tableDirectives": [
      {
        "params": {
          "name": "Order",
          "title": "订单",
        },
        "type": "createTable",
      },
    ],
  },
]
`;

exports[`groupByTable table renamed 1`] = `
[
  {
    "fieldDirectives": [
      {
        "params": {
          "name": "id",
          "notNull": true,
          "primaryKey": true,
          "table": "Product",
          "title": "商品id",
          "type": "Long",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "name",
          "notNull": true,
          "table": "Product",
          "title": "商品名",
          "type": "String",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "price",
          "notNull": true,
          "table": "Good",
          "title": "价格",
          "type": "Decimal",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "description",
          "notNull": true,
          "table": "Good",
          "title": "描述",
          "type": "LongText",
        },
        "type": "addField",
      },
    ],
    "table": "Product",
    "tableDirectives": [
      {
        "params": {
          "name": "Product",
          "newName": "Good",
        },
        "type": "renameTable",
      },
    ],
  },
  {
    "fieldDirectives": [
      {
        "params": {
          "name": "id",
          "notNull": true,
          "primaryKey": true,
          "table": "Order",
          "title": "订单id",
          "type": "Long",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "id",
          "newName": "orderNo",
          "table": "Order",
        },
        "type": "renameField",
      },
      {
        "params": {
          "name": "productId",
          "notNull": true,
          "reference": "Product.id",
          "referenceCardinality": "ManyToOne",
          "table": "Order",
          "title": "商品引用",
          "type": "Reference",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "quantity",
          "notNull": true,
          "table": "Order",
          "title": "数量",
          "type": "Integer",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "totalPrice",
          "notNull": true,
          "table": "Order",
          "title": "总价",
          "type": "Decimal",
        },
        "type": "addField",
      },
      {
        "params": {
          "name": "id",
          "table": "Order",
          "title": "订单编号",
        },
        "type": "updateField",
      },
      {
        "params": {
          "name": "id",
          "table": "Order",
          "type": "String",
        },
        "type": "updateField",
      },
    ],
    "table": "Order",
    "tableDirectives": [
      {
        "params": {
          "name": "Order",
          "title": "订单",
        },
        "type": "createTable",
      },
    ],
  },
]
`;
